src/thermo/PlasmaPhase.cpp

```diff
@@ -0,0 +1,272 @@
+    eedf["energy-levels"] = levels;
+    if (m_distributionType == "isotropic") {
+        eedf["shape-factor"] = m_isotropicShapeFactor;
+        eedf["mean-electron-energy"].setQuantity(meanElectronEnergy(), "eV");
+    } else if (m_distributionType == "discretized") {
+        vector_fp dist(m_nPoints);
+        Eigen::Map<Eigen::ArrayXd>(dist.data(), m_nPoints) = m_electronEnergyDist;
+        eedf["distribution"] = dist;
+        eedf["normalize"] = m_do_normalizeElectronEnergyDist;
+    }
+    phaseNode["electron-energy-distribution"] = std::move(eedf);
+}
+
+void PlasmaPhase::setParameters(const AnyMap& phaseNode, const AnyMap& rootNode)
+{
+    IdealGasPhase::setParameters(phaseNode, rootNode);
+    if (phaseNode.hasKey("electron-energy-distribution")) {
+        const AnyMap eedf = phaseNode["electron-energy-distribution"].as<AnyMap>();
+        m_distributionType = eedf["type"].asString();
+        if (m_distributionType == "isotropic") {
+            if (eedf.hasKey("shape-factor")) {
+                setIsotropicShapeFactor(eedf["shape-factor"].asDouble());
+            } else {
+                throw CanteraError("PlasmaPhase::setParameters",
+                    "isotropic type requires shape-factor key.");
+            }
+            if (eedf.hasKey("mean-electron-energy")) {
+                double energy = eedf.convert("mean-electron-energy", "eV");
+                setMeanElectronEnergy(energy);
+            } else {
+                throw CanteraError("PlasmaPhase::setParameters",
+                    "isotropic type requires electron-temperature key.");
+            }
+            if (eedf.hasKey("energy-levels")) {
+                setElectronEnergyLevels(eedf["energy-levels"].asVector<double>().data(),
+                                        eedf["energy-levels"].asVector<double>().size());
+            }
+            setIsotropicElectronEnergyDistribution();
+        } else if (m_distributionType == "discretized") {
+            if (!eedf.hasKey("energy-levels")) {
+                throw CanteraError("PlasmaPhase::setParameters",
+                    "Cannot find key energy-levels.");
+            }
+            if (!eedf.hasKey("distribution")) {
+                throw CanteraError("PlasmaPhase::setParameters",
+                    "Cannot find key distribution.");
+            }
+            if (eedf.hasKey("normalize")) {
+                enableNormalizeElectronEnergyDist(eedf["normalize"].asBool());
+            }
+            setDiscretizedElectronEnergyDist(eedf["energy-levels"].asVector<double>().data(),
+                                             eedf["distribution"].asVector<double>().data(),
+                                             eedf["energy-levels"].asVector<double>().size());
+        }
+    }
+}
+
+bool PlasmaPhase::addSpecies(shared_ptr<Species> spec)
+{
+    bool added = IdealGasPhase::addSpecies(spec);
+    size_t k = speciesIndex(spec->name);
```

Comment:
Just after a species is added, you know that it's index is `m_kk-1`, so it's not necessary to look it up like this.